### ISSUE TITLE
Async protocol job

### ISSUE THREAD

**Issue by galedric** [atom/electron#410]
Is there a way to return something that is not yet available from a custom protocol handler?

```js
protocol.registerProtocol('atom', function(request) {
  var url = request.url.substr(7)
  return new protocol.RequestAsyncStringJob(function(cb) {
    somewhere.fetch(url, function(data) { cb({ encoding, mimeType, data }); });
  });
});
```

My intent is to lazily fetch data from network if not yet available locally or local cached value is outdated.

**Comment by lealife** [comment]
+1.
I hope this feature will be implemented as soon as possible. Currently, I must start a local server with a port to handle request.

**Comment by steelbrain** [comment]
:+1:

**Comment by mmastrac** [comment]
I think this is fixed in the latest 0.31 series -- protocols use a callback handler instead of requiring an immediate return.

**Comment by zcbenz** [comment]
We still need a way to support passing streams, so we can pipe networking requests or file streams. But current solution seems to satisfy most people's needs, for future needs we can just open another issue.
